File: 0040-combination-sum-ii/0040-combination-sum-ii.java
class Solution {
    List<List<Integer>> res = new ArrayList<>();;
    List<Integer> list = new ArrayList<>();;
    int sum = 0;
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        backtracking(candidates, target, 0);
        return res;
    }

    private void backtracking(int[] candidates, int target, int start) {
        if (sum == target) {
            res.add(new ArrayList<Integer>(list));
            return;
        }

        for (int i = start; i < candidates.length; i++) {
            if (sum + candidates[i] > target) {
                break;
            }

            if (i != start && candidates[i] == candidates[i - 1]) {
                continue;
            }
            list.add(candidates[i]);
            sum += candidates[i];
            backtracking(candidates, target, i + 1);
            sum -= candidates[i];
            list.remove(list.size() - 1);
        }
    }

   
    
}

File: 0040-combination-sum-ii/0040-combination-sum-ii.ts
function combinationSum2(candidates: number[], target: number): number[][] {
    const res : number[][] = [];
    const path : number[] = [];
    let sum : number = 0;
    
    candidates.sort((a, b) => a - b);
    
    function backtrack(start: number) : void {
        if (sum === target) {
            res.push([...path]);
            return;
        }
        
        for (let i = start; i < candidates.length; i++) {
            if (i !== start && candidates[i] === candidates[i - 1]) {
                continue;
            }
            
            if (candidates[i] + sum > target) {
                return;
            }
            
            path.push(candidates[i]);
            sum += candidates[i];
            backtrack(i + 1);
            sum -= candidates[i];
            path.pop();
        }
    }

    backtrack(0);
    return res;
    
};